File: frontend/src/app/services/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class TaskService {
  private apiUrl = 'http://localhost:8080/api/tasks';

  constructor(private http: HttpClient) { }

  // get all tasks
  getTasks(): Observable<any[]> {
    return this.http.get<any[]>(this.apiUrl);
  }

  // get task by id
  getTaskById(id: number): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/${id}`);
  }

  // add a new task
  addTask(task: any): Observable<any> {
    return this.http.post<any>(this.apiUrl, task);
  }

  // update a task
  updateTask(id: number, task: any): Observable<any> {
    return this.http.put<any>(`${this.apiUrl}/${id}`, task);
  }

  // delete a task
  deleteTask(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // delete all tasks
  deleteAllTasks(): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/deleteAll`);
  }

  // search tasks
  searchTasks(searchTerm: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/search`, {
      params: { searchTerm },
    });
  }


  // search tasks by date
  searchTasksByDate(start: string, end: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/searchByDate`, {
      params: { start, end },
    });
  }

  // search tasks by tag
  searchTasksByTag(tag: string): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/searchTasksByTag`, {
      params: { tag },
    });
  }

  // get tags
  getTags(): Observable<string[]> {
    return this.http.get<string[]>(`${this.apiUrl}/tags`);
  }
}

File: backend/src/main/java/com/example/todo_api/service/TaskService.java
package com.example.todo_api.service;

import com.example.todo_api.model.Task;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;

import java.util.ArrayList;
import java.util.List;

@Service
public class TaskService {

    private List<Task> tasks = new ArrayList<>();

    private Long nextId = 1L;

    // method to get all tasks
    public List<Task> getAllTasks() {
        return tasks;
    }

    // method to create a new task
    public Task createTask(Task task) {
        task.setId(nextId++);
        task.setCreatedAt(LocalDateTime.now());
        tasks.add(task);
        return task;
    }

    // method to get a task by id
    public Task getTaskById(Long id) {
        for (Task task : tasks) {
            if (task.getId().equals(id)) {
                return task;
            }
        }
        return null;
    }

    // method to update a task by id
    public Task updateTask(Long id, Task updatedTask) {
        Task existingTask = tasks.stream()
                .filter(task -> task.getId().equals(id))
                .findFirst()
                .orElse(null);

        if (existingTask != null) {
            existingTask.setTitle(updatedTask.getTitle());
            existingTask.setDescription(updatedTask.getDescription());
            existingTask.setTag(updatedTask.getTag());
            existingTask.setCompleted(updatedTask.isCompleted());
            return existingTask;
        }
        return null;
    }

    // method to delete a task by id
    public void deleteTask(Long id) {
        tasks.removeIf(task -> task.getId().equals(id));
    }

    // method to delete all tasks
    public void deleteAllTasks() {
        tasks.clear();
    }

    // method to search by title or description
    public List<Task> searchTasks(String searchTerm) {
        return tasks.stream()
                .filter(task -> task.getTitle().contains(searchTerm) || task.getDescription().contains(searchTerm))
                .toList();
    }

    // method to search by date
    public List<Task> searchTasksByDate(LocalDateTime start, LocalDateTime end) {
        return tasks.stream()
                .filter(task -> task.getCreatedAt().isAfter(start) && task.getCreatedAt().isBefore(end))
                .toList();
    }

    // method to search by tag
    public List<Task> searchTasksByTag(String tag) {
        return tasks.stream()
                .filter(task -> task.getTag().equals(tag))
                .toList();
    }

    public List<String> getTags() {
        return tasks.stream()
                .map(Task::getTag)
                .distinct()
                .toList();
    }
}